### PR TITLE
server: support gzip compression for metrics

### LINKED ISSUE #12355: tikv should support gzip compression for metrics response

### PR DESCRIPTION
**Pull request by glorv**
Signed-off-by: glorv <glorvs@163.com>

### What is changed and how it works?

Issue Number: ref #12355

What's Changed:
Support write metrics response with gzip compression. If the request header contains `Accept-Encoding: gzip`, tikv server will return the response with gzip compressed; If not, still return no-compressed response as before.

In my test in a local environment, use gzip compression can reduce the body size to about 1/12.
```
$ curl -v -i --compressed -s 127.0.0.1:20180/metrics
...
< HTTP/1.1 200 OK
< content-encoding: gzip
< content-type: text/plain; charset=utf-8
< content-length: 118739
< date: Mon, 18 Apr 2022 08:59:22 GMT

$ curl -v -i -s 127.0.0.1:20180/metrics
< HTTP/1.1 200 OK
< content-type: text/plain; charset=utf-8
< content-length: 1430175
< date: Mon, 18 Apr 2022 08:58:24 GMT
```

Performance Impact:
The gzip compression will cost about 12ms extra cpu time, which should not be the bottleneck in a typical 8c environment.

In my benchmark with (1*tidb(40c), 1*pd, 3*tikv(8c,16gb)) with 3k tpcc workload. With a concurrency of 200, each tikv's cpu usage is 700+, compare between v6.0.0 and this commit(based on v6.0.0), there is no visiable performance impact for all of tmpC, P99, P999 metrics.  

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

### Release note 

```release-note
Support encode metrics response with gzip to reduce the http body size
```

### CODE REVIEW

**Comment by BusyJay** [comment]
383k doesn't seem a large cost.

**Comment by glorv** [comment]
> 383k doesn't seem a large cost.

Sorry for the misleading exmple. This example is from a newly bootstraped cluster. For a typical online cluster the body size of the metrics request is about 1.4MiB.  /cc @zhangjinpeng1987 Do you have a brief statictics of how the metrics traffic cost in our cloud environment?

File: Cargo.lock [hunk 1]
```diff
@@ -2720,6 +2721,16 @@ version = "0.2.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "68354c5c6bd36d73ff3feceb05efa59b6acb7626617f4962be322a825e61f79a"
 
+[[package]]
+name = "miniz-sys"
```
Comment (BusyJay) [inline Cargo.lock:2725]: Why this is added?

**Review by BusyJay — COMMENTED** [review]
What's the impact on Prometheus?

File: Cargo.lock (hunk 1 above)
Comment (glorv) [reply Cargo.lock:2725]: It's depended by the [flate2](https://crates.io/crates/flate2) package.

**Comment by glorv** [comment]
> What's the impact on Prometheus?

There is no impact. Prometheus supports and prefers gzip compressed http response.  See: https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format

**Comment by BusyJay** [comment]
> There is no impact. Prometheus supports and prefers gzip compressed http response. See: https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format

Decoding gzip can't have the same performance as reading plain text. I'm worry about the performance when there are a lot of nodes.

**Comment by glorv** [comment]
> > There is no impact. Prometheus supports and prefers gzip compressed http response. See: https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format
> 
> Decoding gzip can't have the same performance as reading plain text. I'm worry about the performance when there are a lot of nodes.

In my benchmark at local environment, for a 1.4MB length metrics content , both gzip encode and decode cost less than 5ms cpu time. Thus, given a cluster with 900 tikv (with is likely to store more than 1PB data), the average metrics qps 900/15 = 60/s, so the extra cpu needed is 5ms * 60 = 300ms. So we only need an extra of 0.3 cores. I think this must not be a bottleneck. I'm not sure if there are some test clusters big enough to check this performance regression.

File: Cargo.lock (hunk 1 above)
Comment (BusyJay) [reply Cargo.lock:2725]: flate2 has been a part of dependency before but miniz-sys is added in this PR.

Comment (glorv) [reply Cargo.lock:2725]: Thanks for the reminding. I change flate2 to use zlib instead of miniz instead.

File: Cargo.toml [hunk 2]
```diff
@@ -90,6 +90,7 @@ engine_traits_tests = { path = "components/engine_traits_tests", default-feature
 error_code = { path = "components/error_code", default-features = false }
 fail = "0.5"
 file_system = { path = "components/file_system", default-features = false }
+flate2 = { version = "=1.0.11", default-features = false, features = ["zlib"] }
```
Comment (BusyJay) [inline Cargo.toml:93]: Why `=`?

File: src/server/status_server/mod.rs [hunk 3]
```diff
@@ -495,6 +502,36 @@ where
         }
     }
 
+    fn handle_get_metrics(req: Request<Body>) -> hyper::Result<Response<Body>> {
+        // the following logic is port from prometheus's golang:
+        // https://github.com/prometheus/client_golang/blob/24172847e35ba46025c49d90b8846b59eb5d9ead/prometheus/promhttp/http.go#L155-L176
+        let encoding = req
```
Comment (BusyJay) [inline src/server/status_server/mod.rs:508]: This may be extracted as a function `client_accept_gzip`.

File: Cargo.toml (hunk 2 above)
Comment (glorv) [reply Cargo.toml:93]: I just copy from `tidb_query_expr` component. Changed to `1.0`.

File: src/server/status_server/mod.rs (hunk 3 above)
Comment (glorv) [reply src/server/status_server/mod.rs]: done

**Comment by glorv** [comment]
@BusyJay @5kbpers  PTAL again, thanks~

File: src/server/status_server/mod.rs [hunk 4]
```diff
@@ -1327,6 +1372,61 @@ mod tests {
...
 
+    #[test]
+    fn test_metrics() {
+        let _test_guard = TEST_PROFILE_MUTEX.lock().unwrap();
+        let mut status_server = StatusServer::new(
+            1,
+            ConfigController::default(),
+            Arc::new(SecurityConfig::default()),
+            MockRouter,
+            std::env::temp_dir(),
```
Comment (BusyJay) [inline src/server/status_server/mod.rs:1383]: Should use `tempfile::TempDir` instead, otherwise they won't be cleaned up.

File: src/storage/config.rs [hunk 5]
```diff
@@ -381,7 +381,7 @@ mod tests {
         let mut cfg = Config::default();
         assert!(cfg.validate().is_ok());
 
-        let max_pool_size = std::cmp::max(1, SysQuota::cpu_cores_quota() as usize);
+        let max_pool_size = std::cmp::max(4, SysQuota::cpu_cores_quota() as usize);
```
Comment (BusyJay) [inline src/storage/config.rs:384]: Why the change?

Comment (glorv) [reply src/storage/config.rs:384]: the original test logic is wrong, change the logic to make the test logic right. 
See: https://github.com/tikv/tikv/blob/1fb8980ccab9ff40c1adc206df52952dab8e8ad8/src/storage/config.rs#L119-L128

File: src/server/status_server/mod.rs (hunk 4 above)
Comment (glorv) [reply src/server/status_server/mod.rs]: done. Replaced all with `tempfile::TempDir`

**Comment by glorv** [comment]
@BusyJay PTAL again, thanks

**Review by BusyJay — APPROVED** [review]

**Review by 5kbpers — APPROVED** [review]
